src/renamer.rs

```diff
@@ -0,0 +1,83 @@
+use crate::matcher::Matcher;
+use crate::name_generator::{NameGenerator, NameGeneratorError};
+use std::error::Error;
+use std::fmt::Display;
+use std::fs::{read_dir, rename};
+use std::path::Path;
+use std::{fmt, io};
+
+#[derive(Debug)]
+pub(crate) enum RenamerError {
+    IoError(io::Error),
+    MatcherError,
+    NameGeneratorError(NameGeneratorError),
+    InvalidFileName,
+}
+
+impl Display for RenamerError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let message = match self {
+            RenamerError::IoError(_) => "An io error occurred",
+            RenamerError::MatcherError => "Invalid input pattern",
+            RenamerError::NameGeneratorError(_) => "Unable to create the new file name",
+            RenamerError::InvalidFileName => "Invalid file name. Make sure it is is valid unicode",
+        };
+
+        write!(f, "{}", message)
+    }
+}
+
+impl Error for RenamerError {}
+
+pub(crate) trait Renamer {
+    fn rename_files_in_directory<P: AsRef<Path>>(&self, directory: P) -> Result<(), RenamerError>;
```

Comment:
Note that this trait is not object-safe due to the generic type param.

Reply:
You're leaking implementation details when using this error.
```suggestion
    fn rename_files_in_directory<P: AsRef<Path>>(&self, directory: P) -> Result<(), Box<dyn Error>>;
```

Reply:
a417b83